Privacy sandbox updates (#2386)
GoogleChrome/developer.chrome.com: site/en/blog/fledge-api/index.md

you can use the thread list to switch between threads, and resume or inspect them more closely as 
appropriate. 

### Observe FLEDGE events

From the Application panel in Chrome DevTools, you can observe FLEDGE interest group and auction
events.

If you visit the FLEDGE demo [shopping site](https://shopping-fledge-demo.glitch.me/advertiser/shopping.html)
in a browser with FLEDGE enabled, information about the `join` event will be displayed.

> {% Img src="image/80mq7dk16vVEg8BBhsVe42n6zn82/3jI5bJh8XKiZP5WHMBYl.png", alt="Screenshot of the

Reviewer: ```suggestion
{% Img src="image/80mq7dk16vVEg8BBhsVe42n6zn82/3jI5bJh8XKiZP5WHMBYl.png", alt="The
```

   DevTools Application panel in Chrome Canary, showing information about a FLEDGE interest group
   join event.", width="800", height="402" %}

Now, if you visit the [FLEDGE demo publisher site](https://publisher-fledge-demo.glitch.me/publisher/index.html?fencedframe)

>    in a browser with FLEDGE enabled, DevTools displays information about the `bid` and `win` events.

Reviewer: ```suggestion
in a browser with FLEDGE enabled, DevTools displays information about the `bid` and `win` events.
```

>    in a browser with FLEDGE enabled, information about the `bid` and `win` events is displayed.

Reviewer: ```suggestion
   in a browser with FLEDGE enabled, DevTools displays information about the `bid` and `win` events.
```

> {% Img src="image/80mq7dk16vVEg8BBhsVe42n6zn82/wMvNrY9GrcD2p3Q6wTsw.png", alt="Screenshot of the

Reviewer: ```suggestion
{% Img src="image/80mq7dk16vVEg8BBhsVe42n6zn82/wMvNrY9GrcD2p3Q6wTsw.png", alt="The
```

   want to add a browser to an interest group for a campaign that isn't running yet, or for some
   other future use, or they may temporarily have run out of advertising budget.

\*\* In the current implementation of FLEDGE, `biddingLogicUrl`, `biddingWasmHelperUrl`,
`dailyUpdateUrl` and `trustedBiddingSignalsUrl` must have the same origin as owner. That may not be
a long-term constraint, and the `ads` and `adComponents` URLs have no such constraint.

{: #update-interest-group}

#### Update interest group attributes

`dailyUpdateUrl` specifies a web server that returns JSON defining interest group properties,

> corresponding to the interest group object passed to `navigator.joinAdInterestGroup()`. This

Reviewer 2: (Applies to line 436 and below -- Github doesn't let me comment there)

The current implementation link can be updated to point to the newer code: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/interest_group_storage.cc;l=671;drc=5a102f146faa0c21eb9cf255ceb46b35a158ab3f

(We could have the code search point to the symbol name, but I'm worried about link breakage if the symbol name changes.)

Also, the following are now supported for updates: priority.

biddingWasmHelperUrl and adComponents probably should be getting updated too (part of the code needed to update them is missing) -- I'll mail a change to make sure they're getting updated too.

Author: Thanks — I've updated the link, and added `biddingWasmHelperUrl ` to the list of attributes that can be changed.

> Also, the following are now supported for updates: priority.

What does this refer to?

Reviewer 2: Actually, we don't support updating biddingWasmHelperUrl or  adComponents, but we should -- that's a bug that we should fix, and then update the documentation after that :)

We do support updating priority.

Author: Got it -- removed `biddingWasmHelperUrl` and `adComponents` for the moment.

I can see `priority` [in the code](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/interest_group_storage.cc;l=671;drc=5a102f146faa0c21eb9cf255ceb46b35a158ab3f) but not in the explainer. Is usage detailed somewhere?

Reviewer 2: @brusshamilton priority is new -- it's used when setting a limit on the number of auction participants in the auction config, via perBuyerGroupLimits, which is a dictionary from buyer origin to the max number of auctions for that buyer -- if a star, '*', is used as a buyer, it sets the default for other unspecified buyers. 

At auction time, we pick a number of interest groups per buyer, according to the perBuyerGroupLimits limit, and we pick the highest priority interest groups (we sort the list by priority, and pick the top n; I think ties are broken arbitrarily).

Reviewer 3: +1 to caraitto's description of priority.
Only thing to clarify is that ties are broken by choosing the interest group with expiration time farther in the future.

Author: Thanks both. 

Will `perBuyerGroupLimits` be documented in the explainer? 

> ties are broken

Excuse my ignorance — what does this refer to?

Reviewer 3: We should probably add it to the explainer. I'll work on putting together a pull request for that.

> > ties are broken
> Excuse my ignorance — what does this refer to?
When we limit the interest groups we pick the highest priority ones first. But if interest groups have the same priority (i.e. they are tied), then we prefer groups with later expiration times.

Author: > ties are broken
>> Excuse my ignorance — what does this refer to?
> > > When we limit the interest groups we pick the highest priority ones first. But if interest groups have the same priority (i.e. they are tied), then we prefer groups with later expiration times.

Ah! Got it. I was imaging some abstruse CS concept :).

Any field not specified in the JSON will not be overwritten—only fields specified in the JSON get
updated—whereas calling `navigator.joinAdInterestGroup()` overwrites any existing interest group.

##### Manual updates

Updates to interest groups owned by the current frame's origin can be triggered manually via
`navigator.updateAdInterestGroups()`. Rate limiting prevents updates from happening too frequently:
repeated calls to `navigator.updateAdInterestGroups()` don't do anything until the rate limit
period (currently one day) has passed. The rate limit gets reset if
`navigator.joinAdInterestGroup()` is called again for the same interest group `owner` and `name`.

##### Automatic updates

> All interest groups loaded for an auction are updated automatically after an auction completes. For

Reviewer 2: Maybe add ", subject to the same rate limits as manual updates."?

Author: Thanks — added.

each owner with at least one interest group participating in an auction, it's as if
`navigator.updateAdInterestGroups()` is called from an iframe whose origin matches that owner.

> Updates are best-effort, and can fail under the following conditions:

Reviewer 2: I might move this so it's clear the best-effort properties apply equally to manually-triggered updates.

Author: Good call — I've moved this up.
